opt: panic in GenerateInvertedJoins with multi-column inverted indexes

Looks like we're calling ExtractRemainingJoinFilters with an empty `rightEq` list in some cases in `GenerateInvertedJoins`.

#### To reproduce:

```
exec-ddl
CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT,
  b INT,
  j JSON,
  INVERTED INDEX (a, j)
)
----

exec-ddl
CREATE TABLE small (
  k INT PRIMARY KEY,
  j JSON
)
----

opt
SELECT small.k, t.k FROM small INNER INVERTED JOIN t ON t.a = 1 AND t.j @> small.j AND t.b = small.k
----
```

#### Stacktrace:

```
    --- FAIL: TestRules/__test (0.00s)
        opt_tester.go:506: testdata/rules/__test:45: runtime error: index out of range [0] with length 0
            (1) assertion failure
            Wraps: (2) attached stack trace
              -- stack trace:
              | github.com/cockroachdb/cockroach/pkg/util/errorutil.ShouldCatch
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/util/errorutil/catch.go:29
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:205
              | runtime.gopanic
              |         /usr/local/opt/go/libexec/src/runtime/panic.go:969
              | runtime.goPanicIndex
              |         /usr/local/opt/go/libexec/src/runtime/panic.go:88
              | github.com/cockroachdb/cockroach/pkg/sql/opt/memo.ExtractRemainingJoinFilters
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/extract.go:265
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*CustomFuncs).GenerateInvertedJoins.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/join_funcs.go:577
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*scanIndexIter).ForEachStartingAfter
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/scan_index_iter.go:258
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*scanIndexIter).ForEach
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/scan_index_iter.go:186
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*CustomFuncs).GenerateInvertedJoins
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/join_funcs.go:488
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreInnerJoin
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:632
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreGroupMember
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:24
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreGroup
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:182
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroup
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:464
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeExpr
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:251
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroupMember
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:506
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroup
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:451
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:225
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).optimizeExpr
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:1645
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).Optimize
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:981
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).RunCommand
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:504
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform_test.runDataDrivenTest.func1.1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer_test.go:234
              | github.com/cockroachdb/datadriven.runDirective.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:321
              | github.com/cockroachdb/datadriven.runDirective
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:326
              | github.com/cockroachdb/datadriven.runDirectiveOrSubTest
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:195
              | github.com/cockroachdb/datadriven.runTestInternal
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:168
              | github.com/cockroachdb/datadriven.RunTest
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:136
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform_test.runDataDrivenTest.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer_test.go:231
              | github.com/cockroachdb/datadriven.Walk
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:413
              | github.com/cockroachdb/datadriven.Walk.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:426
              | testing.tRunner
              |         /usr/local/opt/go/libexec/src/testing/testing.go:1123
              | runtime.goexit
              |         /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1374
            Wraps: (3) runtime error: index out of range [0] with length 0
              | -- cause hidden behind barrier
              | runtime error: index out of range [0] with length 0
              | (1) runtime error: index out of range [0] with length 0
              | Error types: (1) runtime.boundsError
            Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *barriers.barrierError
```

--- a/pkg/sql/opt/memo/expr.go
+++ b/pkg/sql/opt/memo/expr.go
@@ -245,6 +245,28 @@ func (n FiltersExpr) RemoveFiltersItem(search *FiltersItem) FiltersExpr {
 	panic(errors.AssertionFailedf("item to remove is not in the list: %v", search))
 }
 
+// Difference returns a new list filters containing the filters of n that are
+// not in other. If other is empty, n is returned.
+func (n FiltersExpr) Difference(other FiltersExpr) FiltersExpr {
+	if len(other) == 0 {
+		return n
+	}
+	newFilters := make(FiltersExpr, 0, len(n))
+	for i := range n {
+		found := false
+		for j := range other {
+			if n[i].Condition == other[j].Condition {
+				found = true
+				break
+			}
+		}
+		if !found {
+			newFilters = append(newFilters, n[i])
+		}
+	}
+	return newFilters
+}
+
 // RemoveCommonFilters removes the filters found in other from n.
 func (n *FiltersExpr) RemoveCommonFilters(other FiltersExpr) {
 	if len(other) == 0 {

--- a/pkg/sql/opt/memo/expr.go
+++ b/pkg/sql/opt/memo/expr.go
@@ -245,7 +245,7 @@ func (n FiltersExpr) RemoveFiltersItem(search *FiltersItem) FiltersExpr {
 	panic(errors.AssertionFailedf("item to remove is not in the list: %v", search))
 }
 
-// Difference returns a new list filters containing the filters of n that are
+// Difference returns a new list of filters containing the filters of n that are
 // not in other. If other is empty, n is returned.
 func (n FiltersExpr) Difference(other FiltersExpr) FiltersExpr {
 	if len(other) == 0 {
@@ -265,28 +265,6 @@ func (n FiltersExpr) Difference(other FiltersExpr) FiltersExpr {
 		}
 	}
 	return newFilters
-}
-
-// RemoveCommonFilters removes the filters found in other from n.
-func (n *FiltersExpr) RemoveCommonFilters(other FiltersExpr) {
-	if len(other) == 0 {
-		return
-	}
-	// TODO(ridwanmsharif): Faster intersection using a map
-	common := (*n)[:0]
-	for _, filter := range *n {
-		found := false
-		for _, otherFilter := range other {
-			if filter.Condition == otherFilter.Condition {
-				found = true
-				break
-			}
-		}
-		if !found {
-			common = append(common, filter)
-		}
-	}
-	*n = common
 }
 
 // OutputCols returns the set of columns constructed by the Aggregations

--- a/pkg/sql/opt/memo/extract.go
+++ b/pkg/sql/opt/memo/extract.go
@@ -257,6 +257,9 @@ func ExtractRemainingJoinFilters(on FiltersExpr, leftEq, rightEq opt.ColList) Fi
 	if len(leftEq) != len(rightEq) {
 		panic(errors.AssertionFailedf("leftEq and rightEq have different lengths"))
 	}
+	if len(leftEq) == 0 {
+		return on
+	}
 	var newFilters FiltersExpr
 	for i := range on {
 		leftVar, rightVar, ok := isVarEquality(on[i].Condition)

--- a/pkg/sql/opt/memo/extract.go
+++ b/pkg/sql/opt/memo/extract.go
@@ -252,8 +252,11 @@ func ExtractJoinEquality(
 // ExtractRemainingJoinFilters calculates the remaining ON condition after
 // removing equalities that are handled separately. The given function
 // determines if an equality is redundant. The result is empty if there are no
-// remaining conditions.
+// remaining conditions. Panics if leftEq and rightEq are not the same length.
 func ExtractRemainingJoinFilters(on FiltersExpr, leftEq, rightEq opt.ColList) FiltersExpr {
+	if len(leftEq) != len(rightEq) {
+		panic(errors.AssertionFailedf("leftEq and rightEq have different lengths"))
+	}
 	var newFilters FiltersExpr
 	for i := range on {
 		leftVar, rightVar, ok := isVarEquality(on[i].Condition)

--- a/pkg/sql/opt/xform/join_funcs.go
+++ b/pkg/sql/opt/xform/join_funcs.go
@@ -318,9 +318,11 @@ func (c *CustomFuncs) GenerateLookupJoins(
 		// is sufficient.
 		lookupJoin.LookupColsAreTableKey = tableFDs.ColsAreLaxKey(rightSideCols.ToSet())
 
-		// Remove the redundant filters and update the lookup condition.
-		lookupJoin.On = memo.ExtractRemainingJoinFilters(onFilters, lookupJoin.KeyCols, rightSideCols)
-		lookupJoin.On.RemoveCommonFilters(constFilters)
+		// Remove redundant filters from the ON condition if columns were
+		// constrained by equality filters or constant filters.
+		lookupJoin.On = onFilters
+		lookupJoin.On = memo.ExtractRemainingJoinFilters(lookupJoin.On, lookupJoin.KeyCols, rightSideCols)
+		lookupJoin.On = lookupJoin.On.Difference(constFilters)
 		lookupJoin.ConstFilters = constFilters
 
 		if isCovering {
@@ -574,11 +576,9 @@ func (c *CustomFuncs) GenerateInvertedJoins(
 
 		// Remove redundant filters from the ON condition if non-inverted prefix
 		// columns were constrained by equality filters or constant filters.
-		if len(rightSideCols) > 0 || len(constFilters) > 0 {
-			onFilters = memo.ExtractRemainingJoinFilters(onFilters, invertedJoin.PrefixKeyCols, rightSideCols)
-			onFilters = onFilters.Difference(constFilters)
-			invertedJoin.ConstFilters = constFilters
-		}
+		onFilters = memo.ExtractRemainingJoinFilters(onFilters, invertedJoin.PrefixKeyCols, rightSideCols)
+		onFilters = onFilters.Difference(constFilters)
+		invertedJoin.ConstFilters = constFilters
 
 		// Check whether the filter can constrain the inverted column.
 		invertedExpr := invertedidx.TryJoinInvertedIndex(

--- a/pkg/sql/opt/xform/join_funcs.go
+++ b/pkg/sql/opt/xform/join_funcs.go
@@ -480,7 +480,6 @@ func (c *CustomFuncs) GenerateInvertedJoins(
 	eqColsAndOptionalFiltersCalculated := false
 	var leftEqCols opt.ColList
 	var rightEqCols opt.ColList
-	var rightSideCols opt.ColList
 	var optionalFilters memo.FiltersExpr
 
 	var iter scanIndexIter
@@ -523,6 +522,7 @@ func (c *CustomFuncs) GenerateInvertedJoins(
 		// values are joined with the input to create key columns for the
 		// InvertedJoin, similar to GenerateLookupJoins.
 		var constFilters memo.FiltersExpr
+		var rightSideCols opt.ColList
 		for i := 0; i < numPrefixCols; i++ {
 			prefixCol := scanPrivate.Table.IndexColumnID(index, i)
 
@@ -569,13 +569,14 @@ func (c *CustomFuncs) GenerateInvertedJoins(
 			invertedJoin.Input = join
 			invertedJoin.PrefixKeyCols = append(invertedJoin.PrefixKeyCols, constColID)
 			constFilters = append(constFilters, allFilters[allIdx])
+			rightSideCols = append(rightSideCols, prefixCol)
 		}
 
-		// Remove the redundant filters and update the ON condition if there are
-		// non-inverted prefix columns that have been constrained.
+		// Remove redundant filters from the ON condition if non-inverted prefix
+		// columns were constrained by equality filters or constant filters.
 		if len(rightSideCols) > 0 || len(constFilters) > 0 {
 			onFilters = memo.ExtractRemainingJoinFilters(onFilters, invertedJoin.PrefixKeyCols, rightSideCols)
-			onFilters.RemoveCommonFilters(constFilters)
+			onFilters = onFilters.Difference(constFilters)
 			invertedJoin.ConstFilters = constFilters
 		}
 

--- a/pkg/sql/opt/xform/scan_index_iter.go
+++ b/pkg/sql/opt/xform/scan_index_iter.go
@@ -164,6 +164,11 @@ func (it *scanIndexIter) SetOriginalFilters(filters memo.FiltersExpr) {
 // index, the filters are the remaining filters after proving partial index
 // implication (see partialidx.Implicator). Otherwise, the filters are the same
 // filters that were passed to Init.
+//
+// Note that the filters argument CANNOT be mutated in the callback function
+// because these filters are used internally by scanIndexIter for partial index
+// implication while iterating over indexes. In tests the filtersMutateChecker
+// will detect a callback that mutates filters and panic.
 type enumerateIndexFunc func(idx cat.Index, filters memo.FiltersExpr, indexCols opt.ColSet, isCovering bool)
 
 // ForEach calls the given callback function for every index of the Scan

--- a/pkg/sql/opt/xform/testdata/rules/join
+++ b/pkg/sql/opt/xform/testdata/rules/join
@@ -6036,13 +6036,12 @@ project
       │    ├── inverted-expr
       │    │    └── t1.j:20 @> t2.j:3
       │    ├── key: (18)
-      │    ├── fd: (1)-->(2,3), (18)-->(19), (2)==(19), (19)==(2), (1)==(18), (18)==(1)
+      │    ├── fd: (1)-->(2,3), (18)-->(19), (1)==(18), (18)==(1), (2)==(19), (19)==(2)
       │    ├── scan json_arr2 [as=t2]
       │    │    ├── columns: t2.k:1!null l:2 t2.j:3
       │    │    ├── key: (1)
       │    │    └── fd: (1)-->(2,3)
-      │    └── filters
-      │         └── i:19 = l:2 [outer=(2,19), constraints=(/2: (/NULL - ]; /19: (/NULL - ]), fd=(2)==(19), (19)==(2)]
+      │    └── filters (true)
       └── filters
            └── t1.j:8 @> t2.j:3 [outer=(3,8), immutable]
 
@@ -6265,6 +6264,46 @@ anti-join (cross)
  │         └── x:2 IN (1, 3) [outer=(2), constraints=(/2: [/1 - /1] [/3 - /3]; tight)]
  └── filters
       └── y:3 @> column1:1 [outer=(1,3), immutable]
+
+# Regression test for #59859. Do not panic when the ON condition has a constant
+# filter to constrain the non-inverted prefix column and a variable equality
+# filter that does not constrain the prefix column.
+exec-ddl
+CREATE TABLE t59859_a (
+  k INT PRIMARY KEY,
+  j JSON
+)
+----
+
+exec-ddl
+CREATE TABLE t59859_b (
+  k INT PRIMARY KEY,
+  i INT,
+  j JSON,
+  INVERTED INDEX (k, j)
+)
+----
+
+opt expect=GenerateInvertedJoinsFromSelect format=hide-all
+SELECT a.k, b.k
+FROM t59859_a AS a INNER INVERTED JOIN t59859_b AS b
+ON b.k = 1 AND b.j @> a.j AND b.i = a.k
+----
+project
+ └── inner-join (lookup t59859_b [as=b])
+      ├── lookup columns are key
+      ├── inner-join (inverted t59859_b@secondary [as=b])
+      │    ├── flags: force inverted join (into right side)
+      │    ├── inverted-expr
+      │    │    └── b.j @> a.j
+      │    ├── project
+      │    │    ├── scan t59859_a [as=a]
+      │    │    └── projections
+      │    │         └── 1
+      │    └── filters (true)
+      └── filters
+           ├── b.j @> a.j
+           └── i = a.k
 
 # -------------------------------------------------------
 # GenerateInvertedJoinsFromSelect + Partial Indexes

